whelktool/src/main/groovy/datatool/WhelkTool.groovy

```diff
@@ -0,0 +1,476 @@
+package whelk.datatool
+
+import java.util.concurrent.LinkedBlockingDeque
+import java.util.concurrent.ThreadPoolExecutor
+import java.util.concurrent.TimeUnit
+
+import java.sql.ResultSet
+
+import javax.script.ScriptEngineManager
+import javax.script.Bindings
+import javax.script.SimpleBindings
+import javax.script.CompiledScript
+import javax.script.Compilable
+
+import groovy.util.CliBuilder
+import org.codehaus.groovy.jsr223.GroovyScriptEngineImpl
+
+import org.codehaus.jackson.map.ObjectMapper
+
+import whelk.Whelk
+import whelk.Document
+
+
+class WhelkTool {
+
+    static final int DEFAULT_BATCH_SIZE = 500
+    Whelk whelk
+
+    private GroovyScriptEngineImpl engine
+
+    File scriptFile
+    CompiledScript script
+    String scriptJobUri
+
+    String changedIn = "xl"
+    String changedBy = null
+
+    File reportsDir
+    File errorLog
+
+    boolean dryRun
+    boolean noThreads = true
+    boolean stepWise
+    int limit = -1
+
+    private def jsonWriter = new ObjectMapper().writerWithDefaultPrettyPrinter()
+
+    Map<String, Closure> compiledScripts = [:]
+
+    WhelkTool(String scriptPath, File reportsDir=null) {
+        whelk = Whelk.createLoadedCoreWhelk()
+        initScript(scriptPath)
+        this.reportsDir = reportsDir
+        reportsDir.mkdirs()
+        errorLog = new File(reportsDir, "ERRORS.txt")
+
+    }
+
+    private void initScript(String scriptPath) {
+        ScriptEngineManager manager = new ScriptEngineManager()
+        engine = (GroovyScriptEngineImpl) manager.getEngineByName("groovy")
+        scriptFile = new File(scriptPath)
+        String scriptSource = scriptFile.getText("UTF-8")
+        script = ((Compilable) engine).compile(scriptSource)
+        def segment = '/scripts/'
+        def path = scriptFile.toURI().toString()
+        path = path.substring(path.lastIndexOf(segment) + segment.size())
+        scriptJobUri = "https://id.kb.se/generator/scripts/${path}"
+    }
+
+    boolean getUseThreads() { !noThreads && !stepWise }
+
+    Map load(String id) {
+        return whelk.storage.load(id)?.data
+    }
+
+    void selectByIds(Collection<String> ids, Closure process,
+            int batchSize = DEFAULT_BATCH_SIZE) {
+        def uriIdMap = findShortIdsForUris(ids.findAll { it.contains(':') })
+        def shortIds = ids.findResults { it.contains(':') ? uriIdMap[it] : it }
+
+        def idItems = shortIds.collect { "'$it'" }.join(',\n')
+        selectBySqlWhere("id IN ($idItems) AND deleted = false", process,
+                batchSize)
+    }
+
+    Map<String, String> findShortIdsForUris(Collection uris) {
+        def uriIdMap = [:]
+        if (!uris) {
+            return uriIdMap
+        }
+        def uriItems = uris.collect { "'$it'" }.join(',\n')
+        def query = """
+            SELECT id, iri
```

Comment:
Vi behöver väl inte hämta ut iri också (det vi söker på) ?

Reply:
Vi bygger en map av dem sedan (för att behålla ordningen på id:n i in-listan oavsett "typ"), ville inte förlita mig på implicit ordning i `IN (...)` (om det ens går?).